src/diaginterface/diaginterface.jl

```diff
@@ -88,3 +90,21 @@ function setdiagindices!(a::AbstractArray, v, i::Colon)
   diagview(a) .= v
   return a
 end
+
+"""
+    diagonal(v::AbstractVector, [axes]) -> AbstractMatrix
+
+Return a diagonal matrix from a vector `v`, optionally providing the `axes` of the resulting matrix.
+This is an extension of `LinearAlgebra.Diagonal`, designed to avoid the implication of the output type,
+as well as allowing the option of specifying the `axes`.
+Defaults to `Diagonal(v)`.
+"""
+function diagonal(v::AbstractVector)
+  ax = axes(v, 1) # Base.axes1 is private
+  return diagonal(v, (ax, ax))
+end
+function diagonal(v::AbstractVector, ax)
+  vax = axes(v, 1)
+  ax == (vax, vax) || throw(ArgumentError(lazy"Cannot create a `Diagonal` with axes $ax"))
+  return LinearAlgebra.Diagonal(v)
+end
```

Comment:
On second thought, I think we need to think more carefully about this, since we need to decide if `diagonal` should be limited to diagonal matrices or also generalize to arrays. So to keep things simpler let's just go back to the original more narrow definition you had:
```suggestion
diagonal(v::AbstractVector) = LinearAlgebra.Diagonal(v)
```
and think about generalizations later.

Reply:
I don't think I understand what you mean, but I'm fine either way.

Just in case this is what you meant: I do intend that `diagonal` is allowed to return a `Matrix`. To me, usage of `Diagonal` is a performance optimization, which uses a lazy representation of an actual matrix. Sometimes it might be convenient to not have to immediately support that optimization (or it might not even matter at all to have that optimization), `diagonal` could be used instead, meaning "give me some object that is a diagonal matrix, if you have an efficient struct, great, otherwise a regular matrix would also do). In that sense, `LinearAlgebra.diagm` might even be a more correct default implementation here, but I don't think we should think this over too much, considering that an additional benefit of introducing this function is that it becomes incredibly easy to swap out implementations later.

Reply:
I just mean that `diagonal(v::AbstractVector)` is a much more restricted constructor (in a good way), in that it is very clear it is meant to construct some kind of square diagonal matrix where the axes of the row and column match the axes of the input vector. Extending the constructor to `diagonal(v::AbstractVector, ax)` opens up a load of other possibilities, since that API can allow constructing rectangular diagonal matrices, higher dimensional diagonal arrays, etc. `diagonal(v::AbstractVector)` has a clear use case of allowing us to use an alternative to `Diagonal` when there are better square diagonal matrix types we might want to use instead, but we don't have a use case for the more general `diagonal(v::AbstractVector, ax)` yet so we may as well wait to add that since that might determine how we should define it, i.e. a reasonable definition of `diagonal(v::AbstractVector, ax)` might be `DiagonalArray(v, ax)` so it can handle rectangular and higher dimensional diagonal arrays.

I was just thinking through various downsides of the current `diagonal(v::AbstractVector, ax)` definition, since it has a "hard edge" that the API looks like it could allow constructing higher dimensional diagonal arrays and rectangular diagonal matrices but we're kind of arbitrarily restricting it to construct square diagonal matrices with `Diagonal`.